src/System.Memory/src/System/Buffers/Text/Base64Decoder.cs

```diff
@@ -248,52 +292,170 @@ public static OperationStatus DecodeFromUtf8InPlace(Span<byte> buffer, out int b
 
                 if (i0 < 0)
                     goto InvalidExit;
-                WriteThreeLowOrderBytes(ref Unsafe.Add(ref bufferBytes, destIndex), i0);
+
+                WriteThreeLowOrderBytes(ref Unsafe.Add(ref bufferBytes, (IntPtr)destIndex), i0);
                 destIndex += 3;
             }
-            else if (i2 != EncodingPad)
+            else if (t2 != EncodingPad)
             {
-                i2 = Unsafe.Add(ref decodingMap, i2);
+                int i2 = Unsafe.Add(ref decodingMap, (IntPtr)t2);
 
                 i2 <<= 6;
 
                 i0 |= i2;
 
                 if (i0 < 0)
                     goto InvalidExit;
-                Unsafe.Add(ref bufferBytes, destIndex) = (byte)(i0 >> 16);
-                Unsafe.Add(ref bufferBytes, destIndex + 1) = (byte)(i0 >> 8);
+
+                Unsafe.Add(ref bufferBytes, (IntPtr)destIndex) = (byte)(i0 >> 16);
+                Unsafe.Add(ref bufferBytes, (IntPtr)(destIndex + 1)) = (byte)(i0 >> 8);
                 destIndex += 2;
             }
             else
             {
                 if (i0 < 0)
                     goto InvalidExit;
-                Unsafe.Add(ref bufferBytes, destIndex) = (byte)(i0 >> 16);
+
+                Unsafe.Add(ref bufferBytes, (IntPtr)destIndex) = (byte)(i0 >> 16);
                 destIndex += 1;
             }
 
         DoneExit:
-            bytesWritten = destIndex;
+            bytesWritten = (int)destIndex;
             return OperationStatus.Done;
 
         InvalidExit:
-            bytesWritten = destIndex;
+            bytesWritten = (int)destIndex;
             return OperationStatus.InvalidData;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void Avx2Decode(ref byte src, ref byte destBytes, int sourceLength, int destLength, ref uint sourceIndex, ref uint destIndex)
+        {
+            ref byte srcStart = ref src;
+            ref byte destStart = ref destBytes;
+            ref byte simdSrcEnd = ref Unsafe.Add(ref src, (IntPtr)((uint)sourceLength - 45 + 1));
+
+            // The JIT won't hoist these "constants", so help him
+            Vector256<sbyte> lutHi = s_avxDecodeLutHi;
+            Vector256<sbyte> lutLo = s_avxDecodeLutLo;
+            Vector256<sbyte> lutShift = s_avxDecodeLutShift;
+            Vector256<sbyte> mask2F = s_avxDecodeMask2F;
+            Vector256<sbyte> shuffleConstant0 = Vector256.Create(0x01400140).AsSByte();
+            Vector256<short> shuffleConstant1 = Vector256.Create(0x00011000).AsInt16();
+            Vector256<sbyte> shuffleVec = s_avxDecodeShuffleVec;
+            Vector256<int> permuteVec = s_avxDecodePermuteVec;
+
+            //while (remaining >= 45)
+            do
+            {
+                AssertRead<Vector256<sbyte>>(ref src, ref srcStart, sourceLength);
+                Vector256<sbyte> str = Unsafe.As<byte, Vector256<sbyte>>(ref src);
```

Comment:
Only use `Unsafe.As` if this is known to be aligned. Use `Unsafe.LoadUnaligned` otherwise.

Reply:
Here it should not matter, as on x86 unaligned read/writes are emitted, and this section of the code is only executed if SSSE3 or AVX2 is available.

Anyway, I changed it, to make it more obvious that unaligned movs are used.

Reply:
Ups, I missed a point -> https://github.com/dotnet/coreclr/issues/21132
(Forgot to copy over the comment for this). So `Unsafe.As` is the faster option here.

Reply:
BTW: Would it be "more safe" when `Unsafe.As` would emit unaligned read / writes?